core/impl/src/main/java/com/blazebit/persistence/impl/keyset/KeysetManager.java

```diff
@@ -57,15 +68,167 @@ public void initialize(List<OrderByExpression> orderByExpressions) {
         keysetLink.initialize(orderByExpressions);
     }
 
+    public void buildOptimizedKeysetPredicate(StringBuilder sb) {
+        KeysetLink keysetLink = getKeysetLink();
+        KeysetMode keysetMode = keysetLink.getKeysetMode();
+        Keyset keyset = keysetLink.getKeyset();
+        Serializable[] key = keyset.getTuple();
+        OrderByExpression extractedNonNullableExpression = null;
+
+        boolean hasNullableOrderBys = false;
+        for (OrderByExpression orderByExpression : orderByExpressions) {
+            if (orderByExpression.isNullable()) {
+                hasNullableOrderBys = true;
+                break;
+            }
+        }
+
+        extractedNonNullableExpression = orderByExpressions.get(0);
+
+        if (hasNullableOrderBys || !dbmsDialect.supportsRowValueConstructor() || !dbmsDialect.supportsFullRowValueComparison()) {
+            if (extractedNonNullableExpression.isNullable() && (!extractedNonNullableExpression.isNullFirst() || key[0] == null)) {
+                buildKeysetPredicate0(keysetMode, key, sb, orderByExpressions);
+            } else {
+                applyOptimizedKeysetNotNullItem(extractedNonNullableExpression, sb, 0, key[0], keysetMode, orderByExpressions.size() <= 1 && extractedNonNullableExpression.isUnique(), false);
```

Comment:
Extract the condition for an expression being considered `unique` to a variable and document why `orderByExpressions.size() <= 1` is a requirement